### ISSUE TITLE
Change in step selection

### ISSUE THREAD

**Issue by adrienaury** [c4s4/neon#15]
Hello,
Recently upgraded my pipelines to use neon 1.9.1 (upgrade from 1.7.0), i noticed a strange change in step selection and ordering, here is an example to explain what is happening :

```yaml
targets:
  step1:
    steps:
      - print: "step1 depend from nothing"
  step2:
    depends: "step1"
    steps:
      - print: "step2 depend from step1"
  step3:
    depends: ["step1", "step2"]
    steps:
      - print: "step3 depend from step1 and step2"
```

## Neon 1.7.0

The 3 steps are executed each once (since step2 requires step1, and step3 requires step1 and step2, I think this is the correct behavior).

```
❯ neon step3
------------------------------------------------ step1 --
step1 depend from nothing
------------------------------------------------ step2 --
step2 depend from step1
------------------------------------------------ step3 --
step3 depend from step1 and step2
OK
```

## Neon 1.9.1

Step 1 is executed twice !

```
❯ neon step3
------------------------------------------------ step1 --
step1 depend from nothing
------------------------------------------------ step1 --
step1 depend from nothing
------------------------------------------------ step2 --
step2 depend from step1
------------------------------------------------ step3 --
step3 depend from step1 and step2
OK
```

Is it a bug or not ? If not I will have to rewrite a dozen of build.yml files. Thanks for your input !

**Comment by adrienaury** [comment]
# Additionnal infos 

Here is my dependency graph from my build, i left useless relations as comments for information.

```dot
digraph build {
	
    refresh -> info
    // compile -> info
    compile -> refresh
    // lint -> info
    lint -> refresh
    // test -> info
    // test -> refresh
    test -> lint
    // bench -> info
    bench -> refresh
    // release -> info
    // release -> refresh
    // release -> lint
    release -> test
    // "test-int-debug" -> info
    "test-int-debug" -> compile
    // "test-int" -> info
    // "test-int" -> refresh
    // "test-int" -> lint
    // "test-int" -> test
    "test-int" -> bench
    "test-int" -> release
    // publish -> info
    // publish -> refresh
    // publish -> lint
    // publish -> test
    // publish -> release
    publish -> "test-int"
    docker -> info
    // "docker-tag" -> info
    "docker-tag" -> docker
    // "docker-push" -> docker
    "docker-push" -> "docker-tag"
    "docker-push" -> "docker-login"
    // license -> info
    // license -> refresh
    license -> compile

}
```

Even with useless relation disabled, there is some problems with neon 1.9.1, for example when i choose to build the target `test-int`, this happens :

- neon 1.9.1 : info -> refresh -> bench -> info -> refresh -> lint -> test -> release -> test-int
- neon 1.7.0 : info -> refresh -> bench -> lint -> test -> release -> test-int

Neon 1.7.0 correctly sequence targets in the intended order, without executing any target twice. Feels like neon 1.9.1 has an incorrect behavior because some targets are executed multiple times, and therefore increase build time by a lot.

**Comment by c4s4** [comment]
You are perfectly right, the `depends` behavior changed in release _1.7.2_ to fix infinite loop bug. I changed behavior so that if a target is in `depends` list, it runs, even if it already ran before. I though I was the only regular user of this tool, so that this change would only impact me :o)

I suggest these changes:

- Rollback `depends` behavior to as before _1.7.2_.
- Add `after` and `before` options in target definition.

`after` would run given targets before current one (even if they already ran before) and `before` would run targets after current one (even if they already ran before). So that:

```yaml
targets:
  step1:
    steps:
      - print: "step1 depend from nothing"
  step2:
    depends: "step1"
    steps:
      - print: "step2 depend from step1"
  step3:
    depends: ["step1", "step2"]
    steps:
      - print: "step3 depend from step1 and step2"
```

Would print:

```
❯ neon step3
------------------------------------------------ step1 --
step1 depend from nothing
------------------------------------------------ step2 --
step2 depend from step1
------------------------------------------------ step3 --
step3 depend from step1 and step2
OK
```

As you would expect, but:

```yaml
targets:

  step1:
    steps:
    - print: "Step 1"
  
  step2:
    after: [step1]
    steps:
    - print: "Step 2"
  
  step3:
    after: [step1, step2]
    before: [step4]
    steps:
    - print: "Step 3"

  step4:
    steps:
    - print: "Step 4"
```

Would print:

```
❯ neon step3
------------------------------------------------ step1 --
Step 1
------------------------------------------------ step1 --
Step 1
------------------------------------------------ step2 --
Step 2
------------------------------------------------ step3 --
Step 3
------------------------------------------------ step4 --
Step 4
OK
```

What do you think of this proposal?

**Comment by adrienaury** [comment]
If I understand correctly, the 1.7.0 behavior will be back and I can leave all my build.yml as they are so it is ok for me :)

> I though I was the only regular user of this tool, so that this change would only impact me :o)

I think neon is great automation tool because it is technology-agnostic, and can be used everywhere (ci/cd, developper station) and the YAML configuration is very flexible (declarative by default, but can also create some scripting if needed). I didn't find any other tool like this, except format use by specific ci/cd tools (like github actions), but I want to be able to execute my build automation on my dev station

**Comment by c4s4** [comment]
Yes the _1.7.0_ behavior will be back for `depends`. And I will introduce `after` to keep behavior of `1.7.2` and introduce `before` for symmetry reasons. This will be done soon, probably in coming week, but I can't guaranty you any deadline.

**Comment by adrienaury** [comment]
Thank you, no worry for the deadline we can still use neon 1.7.0 for now.

**Comment by c4s4** [comment]
This is fixed in release 1.10.0. Please confirm that this is OK for you.

Regards

**Comment by c4s4** [comment]
I postponed `before` and `after` as this can be done will `call` task in target

**Comment by adrienaury** [comment]
It work thank you
